p2p/src/network/kad/bootstrap/p2p_network_kad_bootstrap_reducer.rs

```diff
@@ -77,31 +93,58 @@ impl P2pNetworkKadBootstrapState {
                                 start: meta.time(),
                             },
                         ));
-                    self.requests.insert(peer_id, request);
+                    state.requests.insert(peer_id, request);
                 }
+
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                let bootstrap_state: &P2pNetworkKadBootstrapState = state.substate()?;
+                let discovery_state: &P2pNetworkKadState = state.substate()?;
+                let key = bootstrap_state.key;
+
+                bootstrap_state
+                    .requests
+                    .iter()
+                    .filter_map(|(peer_id, req)| {
+                        (!discovery_state.requests.contains_key(peer_id)).then_some(
+                            P2pNetworkKadRequestAction::New {
+                                peer_id: *peer_id,
+                                addr: req.addr,
+                                key,
+                            },
+                        )
+                    })
+                    .for_each(|action| dispatcher.push(action));
+
                 Ok(())
             }
             P2pNetworkKadBootstrapAction::RequestDone {
                 peer_id,
                 closest_peers,
             } => {
-                let Some(req) = self.requests.remove(peer_id) else {
-                    return Err(format!("cannot find reques for peer {peer_id}"));
+                let state: &mut P2pNetworkKadBootstrapState = state_context.get_substate_mut()?;
+                let Some(req) = state.requests.remove(peer_id) else {
+                    if state.successful_requests < 20 {
+                        let dispatcher = state_context.into_dispatcher();
```

Comment:
Same as above, unless it complicates the code too much I would rather have the dispatching happening later. Also I see that the dispatching that we see here happens later too. Maybe moving the body of this block (the part that doesn't perform the dispatch) to a helper function will help (so that it can still short-circuit with a return)

Reply:
What I mean is:

```
let result = create_requests_logic(...); // Could even be a method on the state

// Dispatch
let dispatcher = ...;
// .. dispatch logic
```